feat: adds hexcode below the cursor

bun137 opened the pull request:
### Adds the feature to show the hex code under the picker while picking the colour. 

https://github.com/user-attachments/assets/a0c6f83e-fcf4-469e-a9a0-56288f8cd995

![hyprpicker with hexcode img1](https://github.com/user-attachments/assets/8158eeef-6b7a-422b-bbfe-b914e1af3704)

![hyprpicker with hexcode img2](https://github.com/user-attachments/assets/2f77de2b-49df-4495-acc1-fdafbaa78d1d)

fufexan commented:
Looks nice!

One thing I feel is missing is a background for the text. When you hover over white the text is not visible. A background with 0.5 alpha would make it more visible.

danielwerg commented:
Nice! What @fufexan said and is there a reason you put hex code inside the ring? I don't think I have a preference either way but was just wondering.

vaxerski commented:
2/10, should be right below the circle with a background.

bun137 commented:
Thank you! I've added background to the hex (made a simple rectangle behind the text).

I was messing around tryna figure out the best place to put the hex code. Putting it at the bottom of the screen wasn't so great. Plus I had read a comment saying some folks use it to compare colours so kinda made sense to put in the ring!

bun137 commented:
> 2/10, should be right below the circle with a background. Right now it's barely readable and quite annoying

I'll make those changes.

danielwerg commented:
Oh great, no issues functionality-wise anymore for me. For design I would suggest trying to round corners a little, right now it looks a bit bold.

I think this feature should be put behind a flag, I would want to use it 100% of the time but I don't think that going to be the case for most people.

Good point on the hex code position! The only time user won't be able to see is when cursor is at top right corner of the screen or anywhere on X axis at the bottom.

vaxerski commented:
> I think this feature should be put behind a flag, I would want to use it 100% of the time but I don't think that going to be the case for most people.

Yeah, enabled by default though, with a flag to disable

fufexan commented:
![image](https://github.com/user-attachments/assets/a4833a23-cecc-44a3-b3e0-732920810aac)
Looks really nice

fufexan commented:
> Good point on the hex code position! The only time user won't be able to see is when cursor is at top right corner of the screen or anywhere on X axis at the bottom.

Right, this could have a check for the bottom-right corner, and move the text above and left of the cursor.

bun137 commented:
Thanks! Added the rounded corners to the rectangle, I'm working on having a check for the bottom so that I can move the text above the cursor.

danielwerg commented:
I don't know C++ but I did some hard-coding as a PoW
```diff
- cairo_rectangle(PCAIRO, CLICKPOS.x, CLICKPOS.y + 20, 80, 30);

+ if (CLICKPOS.y > (PBUFFER->pixelSize.y - 50) && CLICKPOS.x >(PBUFFER->pixelSize.x - 100)) {
+   cairo_rectangle(PCAIRO, CLICKPOS.x - 80, CLICKPOS.y - 40, 80, 30);
+ } else if (CLICKPOS.y > (PBUFFER->pixelSize.y - 50)) {
+   cairo_rectangle(PCAIRO, CLICKPOS.x, CLICKPOS.y - 40, 80, 30);
+ } else if (CLICKPOS.x > (PBUFFER->pixelSize.x - 100)) {
+   cairo_rectangle(PCAIRO, CLICKPOS.x - 80, CLICKPOS.y + 20, 80, 30);
+ } else {
+   cairo_rectangle(PCAIRO, CLICKPOS.x, CLICKPOS.y + 20, 80, 30);
+ }


- cairo_move_to(PCAIRO, CLICKPOS.x, CLICKPOS.y + 40);

+ if (CLICKPOS.y > (PBUFFER->pixelSize.y - 50) && CLICKPOS.x > (PBUFFER->pixelSize.x - 100)) {
+   cairo_move_to(PCAIRO, CLICKPOS.x - 80, CLICKPOS.y - 20);
+ } else if (CLICKPOS.y > (PBUFFER->pixelSize.y - 50)) {
+   cairo_move_to(PCAIRO, CLICKPOS.x, CLICKPOS.y - 20);
+ } else if (CLICKPOS.x > (PBUFFER->pixelSize.x - 100)) {
+   cairo_rectangle(PCAIRO, CLICKPOS.x - 80, CLICKPOS.y + 40, 80, 30);
+ } else {
+   cairo_move_to(PCAIRO, CLICKPOS.x, CLICKPOS.y + 40);
+ }
```

danielwerg commented:
Add some padding on left for the alpha box, otherwise `#` is touching the border.

bun137 commented:
> I don't know C++ but I did some hard-coding as a PoW
> ```diff
> - cairo_rectangle(PCAIRO, CLICKPOS.x, CLICKPOS.y + 20, 80, 30);
> 
> + if (CLICKPOS.y > (PBUFFER->pixelSize.y - 50) && CLICKPOS.x >(PBUFFER->pixelSize.x - 100)) {
> +   cairo_rectangle(PCAIRO, CLICKPOS.x - 80, CLICKPOS.y - 40, 80, 30);
> + } else if (CLICKPOS.y > (PBUFFER->pixelSize.y - 50)) {
> +   cairo_rectangle(PCAIRO, CLICKPOS.x, CLICKPOS.y - 40, 80, 30);
> + } else if (CLICKPOS.x > (PBUFFER->pixelSize.x - 100)) {
> +   cairo_rectangle(PCAIRO, CLICKPOS.x - 80, CLICKPOS.y + 20, 80, 30);
> + } else {
> +   cairo_rectangle(PCAIRO, CLICKPOS.x, CLICKPOS.y + 20, 80, 30);
> + }
> 
> 
> - cairo_move_to(PCAIRO, CLICKPOS.x, CLICKPOS.y + 40);
> 
> + if (CLICKPOS.y > (PBUFFER->pixelSize.y - 50) && CLICKPOS.x > (PBUFFER->pixelSize.x - 100)) {
> +   cairo_move_to(PCAIRO, CLICKPOS.x - 80, CLICKPOS.y - 20);
> + } else if (CLICKPOS.y > (PBUFFER->pixelSize.y - 50)) {
> +   cairo_move_to(PCAIRO, CLICKPOS.x, CLICKPOS.y - 20);
> + } else if (CLICKPOS.x > (PBUFFER->pixelSize.x - 100)) {
> +   cairo_rectangle(PCAIRO, CLICKPOS.x - 80, CLICKPOS.y + 40, 80, 30);
> + } else {
> +   cairo_move_to(PCAIRO, CLICKPOS.x, CLICKPOS.y + 40);
> + }
> ```

You're right, sorry. My thought process was that the ring size would be same. My bad.

bun137 commented:
I added padding and tested the picker across various resolutions to ensure the hex code is visible everywhere. Also, I updated the hex code to render in uppercase.

bun137 commented:
Here are some visuals!

1920x1080:
https://github.com/user-attachments/assets/e731c5c4-d595-4131-9111-964a56cd2175

640x340:
https://github.com/user-attachments/assets/be2076a6-b028-413e-a705-4f64657c859f

3440x1440:
https://github.com/user-attachments/assets/20eb73bf-7af7-48e5-9a60-584203b05179

danielwerg commented:
Yeah, uppercase hex is better for consistency since that the only format hyprpicker can output right now #55.

Looks perfect to me, need vaxry to review now.

bun137 commented:
@vaxerski could you please review

vaxerski commented on src/hyprpicker.cpp:
```diff
@@ -424,10 +425,58 @@ void CHyprpicker::renderSurface(CLayerSurface* pSurface, bool forceInactive) {
-            cairo_surface_flush(PBUFFER->surface);
+            const auto currentColor = getColorFromPixel(pSurface, CLICKPOS);
+            char       hexBuffer[8];
+            sprintf(hexBuffer, "#%02X%02X%02X", currentColor.r, currentColor.g, currentColor.b);
```
please use std::format

bun137 replied:
okay will do!

bun137 replied:
done!

danielwerg commented:
it still missing a flag to disable this feature, right?

vaxerski commented:
ah, right

bun137 commented:
I've implemented the flag, its `d` to disable live preview of the hex code! Can you check it out?

danielwerg commented on src/main.cpp:
```diff
@@ -15,27 +15,21 @@ static void help(void) {
               << " -t | --no-fractional     | Disable fractional scaling support\n"
-              << " -V | --version           | Print version info\n";
+              << " -V | --version           | Print version info\n"
+              << " -d | --disable-live      | Disable live preview of Hex code\n";
```
+1 for `--disable-hex-preview`. Or maybe even more generic `--disable-preview` to add room for expansion in case hyprpicker supports different formats in future.

bun137 replied:
okay

bun137 commented on src/hyprpicker.cpp:
```diff
@@ -423,11 +424,59 @@ void CHyprpicker::renderSurface(CLayerSurface* pSurface, bool forceInactive) {
+                double padding = 5.0;
+                double textX   = x + padding;
+
+                if (CLICKPOS.y > (PBUFFER->pixelSize.y - 50) && CLICKPOS.x > (PBUFFER->pixelSize.x - 100)) {
```
alright, done

danielwerg reviewed (approved):
`DisableLive` were not renamed internally only in help command, same goes for order of flags. Other than that lgtm.

bun137 commented:
ah sheesh, my bad, will rename the flag

danielwerg commented on src/main.cpp:
```diff
@@ -32,10 +33,11 @@ int main(int argc, char** argv, char** envp) {
                                                {NULL, 0, NULL, 0}};
 
-        int                  c = getopt_long(argc, argv, ":f:hnarzqvtV", long_options, &option_index);
+        int                  c = getopt_long(argc, argv, ":f:hnarzqvtVd", long_options, &option_index);
```
`V` and `d` should be swapped too I guess.

 ```diff
 - int c = getopt_long(argc, argv, ":f:hnarzqvtVd", long_options, &option_index);
 + int c = getopt_long(argc, argv, ":f:hnarzqvtdV", long_options, &option_index);

bun137 replied:
sure, done!

vaxerski reviewed (approved):
lgtm, thanks!

bun137 commented:
This is kinda my first PR to a very popular open source project. Been using Hyprland myself for the last two years and I'm glad I'm able to contribute a little to it! It makes me really happy :)

danielwerg commented:
When should we expect next release with this PR included, is something blocking it?
